### PR TITLE
Update provider-api-server to be deployed in internal and provider mode

### PR DESCRIPTION
**Pull request by rewantsoni**
Allow provider-server to be deployed in intenal and provider mode

Why?
This is a part of RDR for Provider Mode RHSTOR-4886.
With this epic, we are introducing rpc calls for odf to odf communication, the RPC's we have for this epic are limited to Onboarding a Storage Cluster on different Openshift clusters and Getting the Mirroring Info required for setting up RDR (refer #2671).

This can also be used to replace the mechanism we have in the internal mode for setting up Mirroring for blockpool using MCO's Agents.

Hence moving the ocs-provider-server deployment to be deployed in both modes to enable both client to provider communication and cross odf communication.

The PR does the following:
1. Create service, deployment, onboarding job for both modes
2. Update the variable from watchnamespace to podnamespace
3. Remove hardcoded name for storagecluster
4. Move client configmap in storageclient
5. Update the unit tests

### CODE REVIEW

**Comment by nb-ohad** [comment]
@rewantsoni Please provide a proper PR title and description

**Comment by iamniting** [comment]
Can the provider handle requests for multiple storage clusters?

Do we want to run only one ocs-provider-server deployment in case of the multiple storagecluster?

If yes then we can have it as part of the ocs initialization otherwise it should be part of the storagecluster reconciliation.

**Review by iamniting — CHANGES_REQUESTED** [review]
Can you pls add deprication warning in the storage cluster API in first commit?

Also pls separate the generated changes in the 3rd commit and rename the deployment to the `ocs-server`.

**Comment by rewantsoni** [comment]
> Can the provider handle requests for multiple storage clusters?

Yes, the server should be able to handle multiple storage clusters, but the client will onboard to only the storageCluster in the same namespace as the provider-server
> Do we want to run only one ocs-provider-server deployment in case of the multiple storagecluster?

Yes, one provider-server for multiple storageClusters
> If yes then we can have it as part of the ocs initialization otherwise it should be part of the storagecluster reconciliation.

I have added the deployment as a part of csv and secret, service as part of ocsinit

**Review by iamniting — CHANGES_REQUESTED** [review]
Can you pls resolve merge conflict

File: controllers/ocsinitialization/ocsinitialization_controller.go
Comment (iamniting) [inline controllers/ocsinitialization/ocsinitialization_controller.go:705]: lets use r.context

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [inline controllers/storagecluster/provider_server.go:20]: Can you pls add comments why are we deleting it?

File: tools/csv-merger/csv-merger.go
Comment (iamniting) [inline tools/csv-merger/csv-merger.go:7]: Pls fix this import

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/ocsinitialization/ocsinitialization_controller.go
Comment (leelavg) [inline controllers/ocsinitialization/ocsinitialization_controller.go:738]:
```suggestion
		err = r.Client.Create(r.ctx, onboardingSecretGeneratorJob)
```

File: controllers/storagecluster/provider_server.go
Comment (leelavg) [inline controllers/storagecluster/provider_server.go:298]: I see this wasn't moved to ocsinit, what was this secret used for?

File: controllers/storagecluster/storageclient.go
Comment (leelavg) [inline controllers/storagecluster/storageclient.go:88]: I see noobaasub key isn't carried over, intentional?

**Review by leelavg — COMMENTED** [review]
client-op needs an update after this gets merged, isn't it?

File: controllers/storagecluster/provider_server.go
Comment (rewantsoni) [reply controllers/storagecluster/provider_server.go:298]: I didn't find any use for this secret, and I am not sure why we had it. Nitin mentioned that it was used for the old way of getting tokens

File: controllers/storagecluster/storageclient.go
Comment (rewantsoni) [reply controllers/storagecluster/storageclient.go:88]: The other PR was not merged when I raised this PR, thanks for pointing it out adding it

**Comment by rewantsoni** [comment]
> client-op needs an update after this gets merged, isn't it?

@leelavg I don't think we need any update in client o/p after merging this, let me know if I missed anything.

**Comment by leelavg** [comment]
Looked again, yes, no update required.

File: controllers/ocsinitialization/ocsinitialization_controller.go
Comment (iamniting) [inline controllers/ocsinitialization/ocsinitialization_controller.go:633]: // The ServiceType for the OCS API Server is determined based on the following priority:
// 1. The value set in OCSInitialization.
// 2. The value set in InternalStorageCluster.
// This order ensures backward compatibility.

**Review by iamniting — CHANGES_REQUESTED** [review]
Most of it looks good to me Can you please move the ocs server-related code to its own file inside the ocs intiallization controller?

File: controllers/ocsinitialization/ocsinitialization_controller.go
Comment (rewantsoni) [reply controllers/ocsinitialization/ocsinitialization_controller.go:633]: done

File: controllers/util/k8sutil.go
Comment (iamniting) [inline controllers/util/k8sutil.go:48]: Pls add a comment here that this is a ocsintialization CR names.

File: go.mod
Comment (iamniting) [inline go.mod:41]: Pls add these changes in the respective commit itself.

File: controllers/ocsinitialization/ocsserver.go
Comment (iamniting) [inline controllers/ocsinitialization/ocsserver.go:1]: Lets rename this file to ocs_server

**Review by iamniting — CHANGES_REQUESTED** [review]

**Review by leelavg — COMMENTED** [review]
LGTM.

File: controllers/ocsinitialization/ocsserver.go
Comment (rewantsoni) [reply controllers/ocsinitialization/ocsserver.go:1]: done

File: controllers/util/k8sutil.go
Comment (rewantsoni) [reply controllers/util/k8sutil.go]: done

File: go.mod
Comment (rewantsoni) [reply go.mod]: done

File: controllers/ocsinitialization/onboarding.go
Comment (iamniting) [inline controllers/ocsinitialization/onboarding.go:35]: failed to create/ensure job

File: controllers/ocsinitialization/onboarding.go
Comment (iamniting) [inline controllers/ocsinitialization/onboarding.go:26]: // Create the job only if the secret is not found

**Review by iamniting — CHANGES_REQUESTED** [review]

**Comment by obnoxxx** [comment]
@nb-ohad  wrote:

> @rewantsoni Please provide a proper PR title and description

I think the description is leaking too much internal downstream product information to this public project.

Shouldn't we try to avoid this?

**Review by iamniting — APPROVED** [review]
looks good to me.

/hold for @nb-ohad review.

**Comment by nb-ohad** [comment]
/hold 

@rewantsoni 
This PR introduces architecture changes to the way we handle the API service and to the way we are acting on onboarding actions. Looking at the code I think these changes are unnecessary, complicated, and are pushing us in the wrong direction. 
I would like to have more discussion on this

**Comment by nb-ohad** [comment]
Adding to my last comment: 
The API server is not part of the orchestrator, it is part of the application being deployed. Assuming multiple StorageClusters deployed over multiple namespaces, I would expect to see an API server in each of these namespaces, the same way I would expect to see any other orchestrated payload.

We also need to consider that each of these might be configured differently

File: controllers/storagecluster/reconcile.go
Comment (rewantsoni) [inline controllers/storagecluster/reconcile.go:432]: adding it here so that we deploy ocs-provider-server in external mode

File: controllers/storagecluster/provider_server.go
Comment (nb-ohad) [inline controllers/storagecluster/provider_server.go:366]: Why are we removing these?

File: controllers/storagecluster/reconcile.go
Comment (nb-ohad) [reply controllers/storagecluster/reconcile.go:432]: Why do we need to run this in external mode?

File: controllers/util/k8sutil.go
Comment (nb-ohad) [inline controllers/util/k8sutil.go:60]: 1. Why do we need to have an os look on every call? 
2. A look-up into an env var that is mandatory to the operator's operation should not be able to return an error. An error means rescheduling the event. That does not make sense because there is no way for a value to change in the next reconciliation. It is preferable to panic, as an extreme indication that there is a coding error somewhere. 
3. For a mandatory value, an empty string will be considered an errornous value as well


```suggestion
podNamespace := (func() {
	if ns := os.GetEnv(PodNamespaceEnvVar); ns != "" {
		panic fmt.Errorf("%s must be set", PodNamespaceEnvVar)
	} else {
		return ns
	}
})()
func GetPodNamespace() (string) {
	return podNamespace
}
```

File: onboarding-validation-keys-generator/main.go
Comment (nb-ohad) [inline onboarding-validation-keys-generator/main.go:81]: Drop this change in favor of last comment change

File: onboarding-validation-keys-generator/main.go
Comment (nb-ohad) [inline onboarding-validation-keys-generator/main.go:95]: Drop this change in favor of last comment change

File: onboarding-validation-keys-generator/main.go
Comment (nb-ohad) [inline onboarding-validation-keys-generator/main.go:57]:
```suggestion
		klog.Exitf("unexpected number of storageCluster(s) found in %v namespace, expected: 1 actual: %v", operatorNamespace, len(storageClusterList.Items))
	}
	storageCluster := &storageClusterList[0]
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/provider_server.go
Comment (rewantsoni) [reply controllers/storagecluster/provider_server.go:366]: These are not used in the server hence removing the unused code

File: controllers/util/k8sutil.go
Comment (rewantsoni) [reply controllers/util/k8sutil.go:60]: I agree and I am going to raise a separate PR for this, as we have other helper functions similar to this that read the env variable for OperatorNamespace, WatchNamespace

File: controllers/storagecluster/reconcile.go
Comment (rewantsoni) [reply controllers/storagecluster/reconcile.go:432]: I think we decided that provider API server will be running in all modes

File: controllers/storagecluster/storageclient.go
Comment (rchikatw) [inline controllers/storagecluster/storageclient.go:87]: how about using `utils.AddAnnotation` ?

File: controllers/util/k8sutil.go
Comment (rchikatw) [inline controllers/util/k8sutil.go:27]: is there any reason we are not using `WatchNamespaceEnvVar`?

Comment (rewantsoni) [reply controllers/util/k8sutil.go:27]: WatchNamespaceEnvVar is a list of Namespace (from what I recall comma-separated list) the operator should be watching for changes. We don't want the namespace in the onboarding job and provider server to be set as watchNamespace.

We could use OperatorNamespace but this will break if storageCluster is deployed in a different namespace from the operator. Hence using PodNamespace will ensure that these components are deployed in the operator namespace and can get Resources (ex: storageCluster for onboarding job) from the same namespace as it is deployed.

File: controllers/util/k8sutil.go
Comment (nb-ohad) [reply controllers/util/k8sutil.go]: I would prefer we should not start having `os.Getenv` all over the codebase. The point of the util is to abstract the way/location we are getting the pod namespace from and to enforce its existence and  I believe it is the correct approach. 

To solve the issue you presented in your last comment we can decide that the env var is not mandatory for all code paths (which seems to be the case) and remove the panic (leaving you with an empty pod namespace in some cases). Then enforce the existence on a `GetPodNamespace` call. This way the assumption is that if you called `GetPodNamespace` the coder must make sure the env var was set.


The change will look like 
```golang
var podNamespace =  os.GetEnv(PodNamespaceEnvVar)

func GetPodNamespace() (string) {
	if podNamespace != "" {
		panic fmt.Errorf("%s must be set", PodNamespaceEnvVar)
	}
	return podNamespace
}
```

**Comment by nb-ohad** [comment]
LGTM

**Comment by rewantsoni** [comment]
resolved conflicts

**Comment by nb-ohad** [comment]
@iamniting Can you please take a look as well?

File: controllers/storagecluster/storageclient.go
Comment (iamniting) [inline controllers/storagecluster/storageclient.go:11]: We are not using this library in the existing code, We are only using `maps`, pls change this.

**Review by iamniting — CHANGES_REQUESTED** [review]

**Comment by rewantsoni** [comment]
moved unit-tests into a separate commit

**Review by iamniting — APPROVED** [review]

**Pull request merged**
